src/Microsoft.OpenApi/Services/OpenApiWalker.cs

```diff
@@ -222,14 +278,19 @@ internal void Walk(OpenApiOperation operation)
         /// <param name="parameters"></param>
         internal void Walk(IList<OpenApiParameter> parameters)
         {
+            _visitor.Enter(OpenApiConstants.Parameters);
```

Comment:
>_visitor.Enter(OpenApiConstants.Parameters); [](start = 12, length = 44)

I still think that Enter and Exit should be the responsibility of the caller. Here we are assuming that this property will always be called a "parameters", but this can actually be called anything depending on where it is in the doc, and only the caller has that knowledge. This is similar to the case where Walk(tag) doesn't have the knowledge of what the tag should be called, but the caller knows that each tag can be called by an index.

Reply:
I am struggling to think of anywhere in the spec where a list of <something> is named more than one thing.  However, it might happen one day.  My gut was telling me that putting the Enter/Exit inside the walk method was more cohesive, but I will look at it again.

Reply:
I don't think there's anything like that in the spec today. There are places where a non-list is named differently (schema, for example). 

It seems more readable to me if every Walk follows the same pattern, but if you'd rather follow two conventions: one for list and one for non-list, that's also okay to me.

---
In reply to: [163352360](https://github.com/Microsoft/OpenAPI.NET/pull/176#discussion_r163352360) [](ancestors = 163352360)

Reply:
@PerthCharern   I can't bring myself to lift all the Enter/Exit calls out of the Walk methods that they are wrapping.  Seeing them at the start and the end just makes me feel safe.  However, I agree the name of the context should be defined by the caller.  Therefore I'm going to add a parameter to all walkers so that the caller can define the name of the context segment, but the callee actually has the Enter/Exit calls.  This should allow ever Walk method to be consistent whether it is a property or array item.  Hopefully that will make both of us happy :-)

Reply:
I did the changes and I really don't like where it ended up.  I'm going to try something different.